File: 104.maximum-depth-of-binary-tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        isExistLeft = isExistRight = 0
        if root:
            if root.left:
                isExistLeft = self.maxDepth(root.left)
            if root.right:
                isExistRight = self.maxDepth(root.right)
            return max(isExistLeft, isExistRight) + 1
        return 0

# @lc code=end

File: 104.maximum-depth-of-binary-tree.js
/*
 * @lc app=leetcode id=104 lang=javascript
 *
 * [104] Maximum Depth of Binary Tree
 */

// Definition for a binary tree node.
class TreeNode {
  constructor(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
  }
}

// @lc code=start

/**
 * @param {TreeNode} node
 * @param {number} depth
 * @return {number}
 */
const repeat = (node, depth = 0) => {
  if (node === null) {
    return depth;
  }
  return Math.max(repeat(node.left, depth + 1), repeat(node.right, depth + 1));
};

/**
 * @param {TreeNode} root
 * @return {number}
 */
const maxDepth = (root) => {
  return repeat(root);
};
// @lc code=end

